crates/docs/src/validation.rs

```diff
@@ -0,0 +1,70 @@
+use regex::Regex;
+use std::{
+    env, fs, io,
+    path::{Path, PathBuf},
+};
+
+pub fn extract_matches_from_file(
+    file_path: &Path,
+    re: &Regex,
+) -> Result<Vec<String>, std::io::Error> {
+    let content = fs::read_to_string(file_path)?;
+    let matches: Vec<String> = re
+        .captures_iter(&content)
+        .filter_map(|caps| caps.get(1).map(|m| m.as_str().to_string()))
+        .collect();
+
+    Ok(matches)
+}
+
+pub fn extract_matches_from_directory(
+    dir_path: &Path,
+    re: &Regex,
+    extension: Option<&str>,
+) -> io::Result<Vec<String>> {
+    let mut all_snippets = Vec::new();
+
+    for entry in fs::read_dir(dir_path)? {
+        let path = entry?.path();
+
+        if path.is_dir() {
+            all_snippets.extend(extract_matches_from_directory(&path, re, extension)?);
```

Comment:
Maybe let's use this library instead? https://github.com/BurntSushi/walkdir

We already have it in workspace dependencies.